### PR TITLE
Proposal for 0.2: Typeform integration into Django form processing

### PR DESCRIPTION
**Pull request by fsbraun**
Thanks @Jafnee for the first stab at Typeform and Django. I propose as a next step to introduce ``TypeformMixin`` and ``TypeformView`` which allow to seamlessly integrate typeforms into Django's form processing.

For a typeform a Django form is created with all programatic data processing as desired. The ``TypeformMixin`` adds the typeform rendering (either as widget or popup), the ``TypeformView`` the processing of received data. (It replaces Django's ``FormView``). 

@Jafnee : I hope you're still interested in the package. Certainly I would appreciate your thoughts. Also, if you have other priorities right now, pls. let me know.

### CODE REVIEW

**Comment by Jafnee** [comment]
Hi @fsbraun I've had a quick skim through and the idea sounds good to me. I'll have a dive through the PR over the weekend. I don't have any active use cases for this library, but I'm still interested in helping develop it.

Hi @matiboy any chance of having this repo's ownership transfered over to me?

File: README.rst [hunk 1]
```diff
@@ -36,26 +36,21 @@ Add it to your `INSTALLED_APPS`:
...
 
 .. code-block:: html
 
     {% load django_typeform %}
     <html>
         <body>
-            <div class="my-typeform"></div>
-            <script src="{% static 'django_typeform/embed.js' %}"></script>
-            {% typeforms_embed 'https://xxxx.typeform.com/to/xxxxxx' '.my-typeform' '{"hideHeaders": true, "hideFooter": true}' %}
+            {% typeforms_embed 'https://xxxx.typeform.com/to/xxxxxx' 'my-typeform' '{"hideHeaders": true, "hideFooter": true}' %}
```
Comment (Jafnee) [inline README.rst:46]: 👍 Much cleaner

File: django_typeform/forms.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    Extend a Form class to add Typeform funcitonality.
+    """
+    typeform_id = None
+    typeform_url = None
+    typeform_template = 'django_typeform/typeform.html'
+
+    def __init__(self, *args, **kwargs):
+        self.typeform_url = kwargs.pop('typeform_url', self.typeform_url)
+        if self.typeform_url:
+            if self.typeform_url[-1] == '/':						# Skip trailing /
```
Comment (Jafnee) [inline django_typeform/forms.py]: There's some tabs happening here between then `:` and the `#` was this intentional?

File: django_typeform/forms.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                self.typeform_url = self.typeform_url[:-1]
+            self.typeform_id = self.typeform_url.rsplit('/', 1)[-1] # slug is typeform_id
+
+        return super().__init__(*args, **kwargs)
+
+    def typeform_to_query_dict(self, typeform_uid):
+        """Get repsonse from typeform respopnses api as a query dict suitable for POST."""
+        server=TypeformResponses(form=self.typeform_id)
```
Comment (Jafnee) [inline django_typeform/forms.py]:
```suggestion
        server = TypeformResponses(form=self.typeform_id)
```

File: requirements_dev.txt [hunk 4]
```diff
@@ -1,3 +1,6 @@
+django-sekizai
```
Comment (Jafnee) [inline requirements_dev.txt]: Thinking we could add a `-r requirements.txt` here and in requirements_test.txt instead, should make it so we only got 1 spot to manage.

File: django_typeform/forms.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            'sort':         'submitted_at,desc',
+        }, helper_form=self)
+        return result
+
+    def get_uid(self, request):
+        """Generate unique id for each typeform and session"""
+        hash = hashlib.md5(self.typeform_id.encode('utf-8')).hexdigest()
+        if hash in request.session:
+            return request.session[hash]
+        request.session[hash] = get_random_string(31)				# unique id is random
```
Comment (Jafnee) [inline django_typeform/forms.py]: Swap out the tabs here for spaces

File: django_typeform/forms.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        self.typeform_url = kwargs.pop('typeform_url', self.typeform_url)
+        if self.typeform_url:
+            if self.typeform_url[-1] == '/':						# Skip trailing /
+                self.typeform_url = self.typeform_url[:-1]
+            self.typeform_id = self.typeform_url.rsplit('/', 1)[-1] # slug is typeform_id
+
+        return super().__init__(*args, **kwargs)
+
+    def typeform_to_query_dict(self, typeform_uid):
+        """Get repsonse from typeform respopnses api as a query dict suitable for POST."""
```
Comment (Jafnee) [inline django_typeform/forms.py]:
```suggestion
        """Get response from typeform responses api as a query dict suitable for POST."""
```

File: django_typeform/responses.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+if TYPEFORM_SERVER[-1] != '/':  #  add trailing slash if missing
+    TYPEFORM_SERVER += '/'
+TYPEFORM_FORM = getattr(settings, 'TYPEFORM_FORM', None)
+TYPEFORM_TOKEN = getattr(settings, 'TYPEFORM_TOKEN', "")
+TYPEFORM_CHOICES = getattr(settings, 'TYPEFORM_CHOICES', {})
+
+class TypeformResponses(object):
+    form = None
+
+    decode = {
```
Comment (Jafnee) [inline django_typeform/responses.py:41]: Not quite following how `decode` is being used. Could you explain a little bit how and where it's being used

File: tests/settings.py [hunk 8]
```diff
@@ -22,12 +26,30 @@
...
+                'django.template.context_processors.request',
+            ]
+        }
+    },
+]
+
+
 SITE_ID = 1
 
 if django.VERSION >= (1, 10):
```
Comment (Jafnee) [inline tests/settings.py]: We can remove this version check `if` we're only testing against 1.11+

File: django_typeform/forms.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        """Generate unique id for each typeform and session"""
+        hash = hashlib.md5(self.typeform_id.encode('utf-8')).hexdigest()
+        if hash in request.session:
+            return request.session[hash]
+        request.session[hash] = get_random_string(31)				# unique id is random
+        return request.session[hash]
+
+    def render(self, context):
+        """
+        Renders the typeform. Allows to use "{% include typform %}" in templates for
```
Comment (Jafnee) [inline django_typeform/forms.py]: Wrong templatetag example provided here

File: docs/typeformmixin.rst [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+``MyTypeForm`` will be a fully fledged Django form and at the same
+time be linked to the Typeform with given url.
+
+For this link to work, the ``Question reference`` for each Typeform
+question must be set to the matching Django field name, e.g. ``first_name`` for a
+field ``first_name = forms.CharField(max_length=40, blank=True)``. **This requires editiing
+all questions of a typeform and as Typeform warns might break
+other integrations.**
+
+Additionally, the typeform needs a **hidden field** "typeformuid" which
```
Comment (Jafnee) [inline docs/typeformmixin.rst:29]: This looks looks like it's meant to be `typeform_uid`

File: docs/typeformmixin.rst [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+Additionally, it can be rendered as a typeform using include:
+
+.. code-block:: html
+
+    <form method="post">
+      {% include form with mode='popup' %}
+    </form>
+
+This will render a hidden input field with a unique id passed to the
+typeform's hidden ``typeformuid`` field. Upon completion of
```
Comment (Jafnee) [inline docs/typeformmixin.rst:50]: Here too, `typeform_uid`

**Review by Jafnee — CHANGES_REQUESTED** [review]
Left some feedback, also got a question on how `decode` is being used

File: django_typeform/forms.py (hunk 9 above)
Comment (fsbraun) [reply django_typeform/forms.py]: I believe this is correct (except the typo)

File: docs/typeformmixin.rst (hunk 10 above)
Comment (fsbraun) [reply docs/typeformmixin.rst:29]: It turns out that typeform does not allow underscores in field names. I removed them everywhere for typeformuid

File: docs/typeformmixin.rst (hunk 11 above)
Comment (fsbraun) [reply docs/typeformmixin.rst:50]: dito

File: django_typeform/forms.py (hunk 2 above)
Comment (fsbraun) [reply django_typeform/forms.py]: Unintentional...

**Comment by fsbraun** [comment]
Checks pass, Travis just got stuck.

File: django_typeform/responses.py (hunk 7 above)
Comment (fsbraun) [reply django_typeform/responses.py:41]: Did do so. Hope it helps.

**Comment by Jafnee** [comment]
Hey @fsbraun looks good to me. I'll release/publish it this weekend.

Thanks!

**Review by Jafnee — APPROVED** [review]

**Pull request merged**